### ISSUE TITLE
visionOS support

### ISSUE THREAD

**Issue by efroemling** [beeware/Python-Apple-support#207]
I know visionOS itself is pretty new, but I'm curious if this is on the radar? I'd be happy to try and put together a PR myself if not, as this would be fun to play with. I'd hope it wouldn't require a ton of work to adapt the iOS targets?..

**Comment by freakboy3742** [comment]
It's not on my own radar (unless someone wants to donate an Apple Vision to the cause... 😄) 

If someone wants to contribute a patch, I'll take a look at it. I suspect you're right that the amount of work should be minimal - the build process should be largely compatible with the iOS build process. I suspect the biggest complication will be the fixes to CPython itself - you'll need to patch a bunch of things (mostly in configure, but a few other places as well) to accomodate visionOS as a build target.

**Comment by sheikheddy** [comment]
Hey, I have an Apple Vision Pro and have some experience with CPython, but I'm completely new to BeeWare (only found it via PEP 730) so wouldn't be comfortable committing to writing a patch to add visionOS 2.0 support.

My understanding is that visionOS is a fork of iPadOS, so I'm happy to clone this repo and swap some names in config and using ChatGPT to work through initial issues to test whether `make visionOS` works. 

Would you suggest doing that with 3.13.0 candidate 2, or waiting till 3.13.0 final?

**Comment by freakboy3742** [comment]
My understanding is also that visionOS is essentially "iPadOS on different hardware", so making modifications to support visionOS is likely not *that* complicated.

There are two sets of changes that are required. 

The first is changes to the makefile in this repo to add visionOS support. That's mostly a case of "add visionOS to the list of tags"; the hardest part is working out exactly what Apple "calls" visionOS and the visionOS simulator (there's some weird inconsistencies - it's `iphoneos`, but then `iphonesimulator`; `appletvos` but then `watchos`).

The second is to modify the patch to the CPython sources. The patch that is applied to the CPython sources is currently based on 3.13.0rc2; the CPython core team decided that there will be an rc3 due to a couple of last minute issues that have been discovered, but those shouldn't affect anything on the Apple mobile platforms. The patch is generated using [the 3.13-patched branch](https://github.com/freakboy3742/cpython/tree/3.13-patched) from my fork of the CPython repo; it's literally the 3.13.0rc2 sources, plus 2 patches for adding tvOS and watchOS support. If you wanted to try adding an additional patch for visionOS support, it will mostly be a case of "everywhere the patch says `"tvOS", "watchOS"`, add `"visionOS"`", then re-running `autoconf` to get an updated configure script.

If you need any further pointers, let me know; I'm happy to help get someone up to speed to work on this.

**Comment by sheikheddy** [comment]
Thanks for the thorough response! 

1) The first part sounds feasible: I can grep for all the places with lists of tags, and add ones for visionOS and the visionOS simulator. I'll try my best to get all of them, but if I don't succeed, that can be caught at review time. I read a news article which said that the old name for visionOS was `xros`, but I'm not sure if that is still in use. I can confirm this experimentally or ask around in a discord with visionOS devs.

2) I wasn't aware that there was going to be an rc3. I took 5 minutes to check if this information was public, but it doesn't seem to be. Hopefully someone is assigned to update PEP 719, but it's alright if that's low priority.

3) I can add a release.visionOS.exclude copying the iOS one (i.e, only exclude __pycache__). I don't fully understand what is going on in sitecustomize, but I think I'll need to create a sitecustomize.visionOS.py and include it in my PYTHONPATH. I read the blog post linked, it didn't seem to imply there was anything additional I'd need to change, though I can try and find out.

4) What options would I need to pass to `autoconf`? Are they the same on all runs, or different in the first and subsequent runs? I googled and read a tutorial, but did not find an existing `configure.ac` file to use.
- I had a look at /patch/Python/Python.patch, under "How do I build Python for watchOS" (lines 1592 to 1596) https://github.com/beeware/Python-Apple-support/blob/d8fdaaec55b845e6e3bfcee9439d4a462c0ab9a6/patch/Python/Python.patch#L1592-L1596. This references a ./configure command, though I don't have more context.
- For `--with-build-python` I can use the 3.13 patched branch you shared, and specify some arbitrary path for `--with-framework-name` to place the Python.framework build for visionOS.

Does that sound okay so far? Let me know if I'm on the right track.

**Comment by freakboy3742** [comment]
> 1. The first part sounds feasible: I can grep for all the places with lists of tags, and add ones for visionOS and the visionOS simulator. I'll try my best to get all of them, but if I don't succeed, that can be caught at review time. I read a news article which said that the old name for visionOS was `xros`, but I'm not sure if that is still in use. I can confirm this experimentally or ask around in a discord with visionOS devs.

The way to check this is to see how `xcrun —show-sdk-path —sdk <visionOS?>` responds. If you get a path that is clearly a Vision OS SDK, you’re on the right track.

> 2. I wasn't aware that there was going to be an rc3. I took 5 minutes to check if this information was public, but it doesn't seem to be. Hopefully someone is assigned to update PEP 719, but it's alright if that's low priority.

The RC3 decision that was made late Friday afternoon as the core team sprint was wrapping up. I imagine PEP 719 will be updated as soon as everyone returns from their various flights. I believe the plan is to cut the release on Monday.

> 3. I can add a release.visionOS.exclude copying the iOS one (i.e, only exclude `__pycache__`). I don't fully understand what is going on in sitecustomize, but I think I'll need to create a sitecustomize.visionOS.py and include it in my PYTHONPATH. I read the blog post linked, it didn't seem to imply there was anything additional I'd need to change, though I can try and find out.

IIRC, the sitecustomize file is the same on all mobile platforms - if it’s isn’t, the changes are minimal. It’s an implementation of a trick that fools pip into thinking it’s running *on* the device; it’s needed until pip is able to support a platform - https://github.com/pypa/pip/pull/12962 will do this for iOS, but the sitecustomize will be needed for tvOS, watchOS and visionOS until they become Tier 3 Python platforms.

> 4. What options would I need to pass to `autoconf`? Are they the same on all runs, or different in the first and subsequent runs? I googled and read a tutorial, but did not find an existing `configure.ac` file to use.
 
autoconf has no options - it’s a script that compiles configure.ac into configure. The only thing you’ll need to watch is that you have the right version of autoconf installed - as of Python 3.13, you need to have 2.71. You only need to invoke autoconf if you modify configure.ac… but you’ll need to do that to add a new platform.

> * I had a look at /patch/Python/Python.patch, under "How do I build Python for watchOS" ...
>   . This references a ./configure command, though I don't have more context.

There an analogous guide for iOS and tvOS as well; the iOS one (which is in 3.13 proper) will be the most reliable guide. Those READMEs are essentially the long form documentation version of what Python-Apple-support encodes as a Makefile.

> * For `--with-build-python` I can use the 3.13 patched branch you shared, and specify some arbitrary path for `--with-framework-name` to place the Python.framework build for visionOS.

You can use any recent 3.13 release as the build python - the official rc2 binaries are fine (and are what the Makefile in this repo defaults to). 

Adding a new OS to the list on L36 of the makefile, plus the corresponding TARGETS- and VERSION_MIN- definitions should be all you need to get the Python-Apple-support Makefile to trigger a build for that platform. The only question then is whether there is a need for any additional autoconf overrides.

> Does that sound okay so far? Let me know if I'm on the right track.

Broadly, yes; however, my first suggestion would be to get familiar with how configure.ac, configure, and Makefile work in the CPython repo itself, on a “vanilla” macOS build. It’s a lot easier to work out what is going on when you don’t have cross-platform issues to deal with; but once you’ve got those fundamentals in place, fitting cross-compilation into the picture will be easier. The CPython dev guide (https://devguide.python.org) may be helpful here.

**Comment by rmartin16** [comment]
> > 2\. I wasn't aware that there was going to be an rc3. I took 5 minutes to check if this information was public, but it doesn't seem to be. Hopefully someone is assigned to update PEP 719, but it's alright if that's low priority.
> 
> The RC3 decision that was made late Friday afternoon as the core team sprint was wrapping up. I imagine PEP 719 will be updated as soon as everyone returns from their various flights. I believe the plan is to cut the release on Monday.

FYI: Thomas made an [announcement](https://discuss.python.org/t/incremental-gc-and-pushing-back-the-3-13-0-release/65285) on Discourse.

**Comment by sheikheddy** [comment]
@rmartin16 Thanks for the link to Discourse. I made an account and just caught up on the historical [PEP 730 discussions](https://discuss.python.org/t/pep-730-adding-ios-as-a-supported-platform/35854/39) from last year, which was very helpful reading (even though some of it may be out of date by now due to more recent changes).

@freakboy3742 I appreciate the suggestion to check out the devguide! The background I was looking for was at https://devguide.python.org/getting-started/setup-building/#regenerate-configure.

**Comment by adamroszyk** [comment]
Hey just wanted to ask if you were able to make progress with making it compatible with VisionOS?
I'll be happy to help with testing on device.
Thanks!

**Comment by freakboy3742** [comment]
No - there has been no progress on Vision OS support. We're certainly open to accepting patches if anyone wants to contribute them (see #117 for links to recent tickets showing what is required to add a new platform); but adding VisionOS support isn't a priority for the core team at present.

**Comment by sheikheddy** [comment]
Yeah, this is on me for offering to work on it and then going radio silent for half a year without authoring changes, sorry about that. 

I don't NEED the python support, so it hasn't been a priority, but I still have my notes and am eager to hear your use case. It's something I'm planning to get around to doing but putting it at the top of my list would require demand that isn't really there.

**Comment by johnzhou721** [comment]
Hi! I can try to work this. I already got the patch done, but I have a different version of aclocal (or probably some other version of those packages along the lines) and it keeps adding code checking for pkg-config which results in an error since it checks for platform-triplet-pkg-config or smth like that...it also happens when I just run autoreconf on a brand new python tarball... 

Pray I don't make more typoes... this probably isn't even close to done

**Comment by sheikheddy** [comment]
I will probably be too busy to help until July, @freakboy3742 could you take a look?

**Comment by freakboy3742** [comment]
@johnzhou721 I've told you this several times now - but I get a notification on every pull request, comment, and issue in BeeWare repositories. You don't need to add a comment to notify me that there's a comment in another repository unless it's been a couple of days and it looks like I might have missed that you were looking for feedback on something (and I do mean *days* - I will, from time to time, have a weekend away from the keyboard, or be on a plane flying to a conference, or have other  work priorities that require my attention).

**Comment by johnzhou721** [comment]
FWIW -- https://github.com/johnzhou721/PyDE is a minimum working example for actually embedding the framework into an app, that will work after #283 is merged.

A working picture in a simulator of some trivial Python code being ran:

![Image](https://github.com/user-attachments/assets/729af890-0653-4931-8c41-e82e79426e94)

**Comment by johnzhou721** [comment]
@adamroszyk If you're still hanging around: I've added support for visionOS, and since you said you can test on-device, you can build the branch from ``main`` (``make visionOS``), and there'll be a testbed directory. Open up that Xcode project, disable all the signal handlers (you can see the iOS/README in the official cpython repo for that), and test (using the Xcode testing infrastructure) it for device.

If you want to try to write some Python code that only uses stdout on visionOS to experiment w/ some actual code you might have: see my link above, but wait til the PR mentioned is merged. And also for anyone else -- that project does not save your code.
